components/de_amp/browser/de_amp_url_loader.cc

```diff
@@ -0,0 +1,399 @@
+void DeAmpURLLoader::OnReceiveCachedMetadata(mojo_base::BigBuffer data) {
+  destination_url_loader_client_->OnReceiveCachedMetadata(std::move(data));
+}
+
+void DeAmpURLLoader::OnTransferSizeUpdated(int32_t transfer_size_diff) {
+  destination_url_loader_client_->OnTransferSizeUpdated(transfer_size_diff);
+}
+
+void DeAmpURLLoader::OnStartLoadingResponseBody(
+    mojo::ScopedDataPipeConsumerHandle body) {
+  VLOG(2) << __func__ << " " << response_url_;
+  state_ = State::kLoading;
+  body_consumer_handle_ = std::move(body);
+  body_consumer_watcher_.Watch(
+      body_consumer_handle_.get(),
+      MOJO_HANDLE_SIGNAL_READABLE | MOJO_HANDLE_SIGNAL_PEER_CLOSED,
+      base::BindRepeating(&DeAmpURLLoader::OnBodyReadable,
+                          base::Unretained(this)));
+  body_consumer_watcher_.ArmOrNotify();
+}
+
+void DeAmpURLLoader::OnComplete(
+    const network::URLLoaderCompletionStatus& status) {
+  DCHECK(!complete_status_.has_value());
+  switch (state_) {
+    case State::kWaitForBody:
+      // An error occured before receiving any data.
+      DCHECK_NE(net::OK, status.error_code);
+      state_ = State::kCompleted;
+      if (!throttle_) {
+        Abort();
+        return;
+      }
+      throttle_->Resume();
+      destination_url_loader_client_->OnComplete(status);
+      return;
+    case State::kLoading:
+    case State::kSending:
+      complete_status_ = status;
+      return;
+    case State::kCompleted:
+      destination_url_loader_client_->OnComplete(status);
+      return;
+    case State::kAborted:
+      NOTREACHED();
+      return;
+  }
+  NOTREACHED();
+}
+
+void DeAmpURLLoader::FollowRedirect(
+    const std::vector<std::string>& removed_headers,
+    const net::HttpRequestHeaders& modified_headers,
+    const net::HttpRequestHeaders& modified_cors_exempt_headers,
+    const absl::optional<GURL>& new_url) {
+  // DeAmpURLLoader starts handling the request after
+  // OnReceivedResponse(). A redirect response is not expected.
+  NOTREACHED();
+}
+
+void DeAmpURLLoader::SetPriority(net::RequestPriority priority,
+                                 int32_t intra_priority_value) {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->SetPriority(priority, intra_priority_value);
+}
+
+void DeAmpURLLoader::PauseReadingBodyFromNet() {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->PauseReadingBodyFromNet();
+}
+
+void DeAmpURLLoader::ResumeReadingBodyFromNet() {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->ResumeReadingBodyFromNet();
+}
+
+void DeAmpURLLoader::OnBodyReadable(MojoResult) {
+  if (state_ == State::kSending) {
+    // The pipe becoming readable when kSending means all buffered body has
+    // already been sent.
+    ForwardBodyToClient();
+    return;
+  }
+
+  size_t start_size = buffered_body_.size();
+  uint32_t read_bytes = kReadBufferSize;
+  buffered_body_.resize(start_size + read_bytes);
+  auto result = body_consumer_handle_->ReadData(
```

Comment:
Can we get call OnBodyReadable() when `buffered_body_` isn't empty?
If no why not `start_size = 0;` ?
If yes why do we ask for extra `kReadBufferSize` bytes, not the remaining?